lib/Doctrine/ORM/UnitOfWork.php

```diff
@@ -3329,7 +3329,17 @@ private function isIdentifierEquals($entity1, $entity2)
             ? $this->entityIdentifiers[$oid2]
             : $this->identifierFlattener->flattenIdentifier($class, $class->getIdentifierValues($entity2));
 
-        return $id1 === $id2 || implode(' ', $id1) === implode(' ', $id2);
+        return $id1 === $id2 || $this->hashIdentifier($id1) === $this->hashIdentifier($id2);
+    }
+
+    /**
+     * @param array $identifier
+     *
+     * @return string
+     */
+    private function hashIdentifier(array $identifier)
+    {
+        return serialize(array_values(array_map('strval', $identifier)));
```

Comment:
Well I don't have a strong opinion on this. I tend to prefer functionality also considering that the whole communication process between Doctrine and the DB takes a lot of more time than having that serialization here. Sometimes I find that optimization Doctrine wise over exaggerated because if the retrieval from DB will always be much slower than PHP code evaluation (especially if you do not have a blazing fast DB server). Also considering that PHP7 will most probably be a lot faster, I don't see any problems with this. Furthermore I think that people having those edge cases will surely be willing to take the minimal performance difference. But that is just my 2 cents. In the end if you want to query a lot of entities you might nevertheless be better off doing pure DBAL queries. Hope that helps...